pkg/common/ipsec/utils.go

```diff
@@ -17,13 +20,23 @@ const (
 
 func CountUniqueIPsecKeys(states []netlink.XfrmState) int {
 	keys := make(map[string]bool)
+	invalidStateFound := false
 	for _, s := range states {
-		if s.Aead == nil {
+		if s.Aead != nil && s.Auth == nil && s.Crypt == nil {
+			keys[string(s.Aead.Key)] = true
+			continue
+		}
+		if s.Aead == nil && s.Auth != nil && s.Crypt != nil {
+			// we want to count the number of unique (Auth, Crypt) tuples
+			key := fmt.Sprintf("%s:%s", string(s.Auth.Key), string(s.Crypt.Key))
+			keys[key] = true
 			continue
 		}
-		keys[string(s.Aead.Key)] = true
+		invalidStateFound = true
+	}
+	if invalidStateFound {
+		fmt.Fprintf(os.Stderr, "an unsupported XfrmStateAlgo combination has been found\n")
```

Comment:
Since we know that this is an invalid state, why not also return an error in this case?

Reply:
That was the initial idea but it causes regression: https://github.com/cilium/cilium/pull/29182#pullrequestreview-1741248733

Reply:
Is there a way for us to mark XFRM states as being owned by Cilium?

Reply:
I have no idea... (((

Reply:
Not really, no, and it feels fairly out-of-scope for this PR anyway.

Reply:
Feel free to send a PR or CFP with a proposed approach. The goal of this PR is to add support for non-AEAD keys, not to deal with concurrent software using XFRM.

Reply:
@asauber @derailed I'm going to add error to the function signature.
Are you okay if the function returns the IPsec key count even if an error was detected? or it should return 0?
Thanks!

Reply:
Non-zero for other parameters is definitely fine

Reply:
Done.